### ISSUE TITLE
Zero Downtime Deployments

### ISSUE THREAD

**Issue by divrse** [caprover/caprover#661]
**Is your feature request related to a problem? Please describe.**
Zero downtime deployments are a must for long running applications that always have connected users.
I am a big fan of cap rover and this is it's only lacking area in my opinion.

**Describe the solution you'd like**
Almost all deployment systems have zero downtime deployments, even dokku does.
What are the technical limitations to making this a reality for caprover?
Obviously in your head it's as easy as keeping the previous replica alive till new one is active, or having multiple replicas and then removing the old ones when deployment is active, but since this feature has not existed for years now, it for sure is a bit more complicated than that.
I would be happy to contribute development to this feature if I can.

**Describe alternatives you've considered**
Not sure.

**Additional context**
This would really make caprover a lot better and more salable.

**Comment by githubsaturn** [comment]
TLDR:
CapRover does support zero downtime (mostly). What makes you think that CapRover doesn't have zero downtime?

More details:
If you're app doesn't use a volume, CapRover uses `start-first` strategy when updating the containers. This means that the newer version of your container will be up and running before the old one is killed. This should give you next-to-zero downtime. There is still a slight chance that your container wakes up but internally it takes some time before it can accept requests. This is out of CapRover's hands.

This strategy is intentionally not applied to apps with volumes attached to them. This is because if multiple instances of the same service are trying to access the same file, this will result in data corruption and failures. For apps with volumes (persistent data), CapRover user `stop-first` strategy. This means that the old container is stopped first, before the new container starts up. This results in a few seconds of downtime.

**Comment by slava-vishnyakov** [comment]
> There is still a slight chance that your container wakes up but internally it takes some time before it can accept requests.

For us, it's 100% of the time and users end up seeing 502 CapRover screen instead of the app. But it wasn't this way when we used Dokku. We like CapRover way more, but absence of real zero-downtime deployments is a real mood killer.

> This is out of CapRover's hands.

Actually, it's not quite that. Dokku has it. 

If you look at Dokku https://dokku.com/docs/deployment/zero-downtime-deploys/ 
scroll down to "Check instructions", basically, you can create a simple file called `CHECKS` file, like this:

```
/about  Our Amazing Team
/stylesheets/index.css  .body
```

After a deploy and starting the contained Dokku will wait for `DOKKU_CHECKS_WAIT` (default: 5) seconds, then try to `curl http://{container-ip}:{active-port}/about` and see if `Our Amazing Team` is there, if it is it will proceed to `curl http://{container-ip}:{active-port}/stylesheets/index.css` and check if `.body` (string) is there. If it is, it will switch nginx from the old container to the new one. If it doesn't find these strings within `DOKKU_CHECKS_TIMEOUT` (default: 30) seconds, it will restart tests up to `DOKKU_CHECKS_ATTEMPTS` (default: 5) times, and if they fail, it will mark deploy as failed and reject it.

These also act as smoke tests and saved us numerous times from bad deploys. But they also ensure that containers are really ready to accept traffic. I can probably devote some time to try to implement it in CapRover, but maybe you can give me some hints as how this could be implemented?

**Comment by githubsaturn** [comment]
@slava-vishnyakov  - there is a different implementation across Dokku and CapRover. Dokku has to manage the lifecycle of containers itself since it doesn't use an orchestration system as it was created way before Docker swarm. CapRover on the other hand, uses Docker swarm.

But that doesn't mean that you should see 502 page during deploys. Docker swarm is supposed to handle all checks automatically through HEALTHCHECK.

- Is your app stateless or does it have any mounted volumes?
- Are you using a specific HEALTHCHECK in your dockerfile?

**Comment by frederikhors** [comment]
My deploy happen every hour or so during the day with no issue at all with Caprover!

**Comment by slava-vishnyakov** [comment]
>  there is a different implementation across Dokku and CapRover

@githubsaturn I understand, thank you!

> * Is your app stateless or does it have any mounted volumes?

Stateless.

> * Are you using a specific HEALTHCHECK in your dockerfile?

We used to have HEALTHCHECKs, but the problem is that they aren't "deploy" checks, they are "healthchecks", so they keep banging our container non-stop. And if there's a temporary hiccup, it starts a cascade of restarts (in our case). I guess we have no option but to implement some kind of healthcheck that would simulate a deploy check.

The bad part about this is that we're both stateless and we don't use Docker Swarm, so a deploy check would be really helpful here. (Though, I think that Dokku even does zero-downtime deploys for stateful containers)

Anyway, thank you for taking the time to explain it.

@frederikhors There are no problems with deploys on caprover, the problem is that our container takes 5-30 seconds before it's really "responsive" due to how it works, but CapRover makes it "live" immediately. So, the users see 502 for quite some time.

**Comment by slava-vishnyakov** [comment]
Oh and I forgot to mention one more benefit that Dokkus zero-downtime deploys have over `HEALTHCHECKS` in CapRover. If you deploy and your HEALTHCHECK for some reason fails, you see that you deployed successfull, but you actually see the old version (so, in a sense your deploy failed) without any notification that your deploy failed. 

I'm not sure how to solve this in CapRover (with the Swarm and all), so I'm just thinking out loud.

**Comment by githubsaturn** [comment]
> We used to have HEALTHCHECKs, but the problem is that they aren't "deploy" checks, they are "healthchecks", so they keep banging our container non-stop. And if there's a temporary hiccup, it starts a cascade of restarts (in our case). I guess we have no option but to implement some kind of healthcheck that would simulate a deploy check.

The HEALTHCHECKs are supposed to be very light and just really report the health of the application. You can define delays, attempts etc to match the behavior you need.


Search for "health" to see all tweaks that you can make to your service to make healthcheck work for your use case. 

I will be very surprised if there is a common usecase that is not covered by Docker's HEALTHCHECKs.

**Comment by Smyja** [comment]
> > There is still a slight chance that your container wakes up but internally it takes some time before it can accept requests.
> 
> For us, it's 100% of the time and users end up seeing 502 CapRover screen instead of the app. But it wasn't this way when we used Dokku. We like CapRover way more, but absence of real zero-downtime deployments is a real mood killer.
> 
> > This is out of CapRover's hands.
> 
> Actually, it's not quite that. Dokku has it.
> 
> If you look at Dokku https://dokku.com/docs/deployment/zero-downtime-deploys/ scroll down to "Check instructions", basically, you can create a simple file called `CHECKS` file, like this:
> 
> ```
> /about  Our Amazing Team
> /stylesheets/index.css  .body
> ```
> 
> After a deploy and starting the contained Dokku will wait for `DOKKU_CHECKS_WAIT` (default: 5) seconds, then try to `curl http://{container-ip}:{active-port}/about` and see if `Our Amazing Team` is there, if it is it will proceed to `curl http://{container-ip}:{active-port}/stylesheets/index.css` and check if `.body` (string) is there. If it is, it will switch nginx from the old container to the new one. If it doesn't find these strings within `DOKKU_CHECKS_TIMEOUT` (default: 30) seconds, it will restart tests up to `DOKKU_CHECKS_ATTEMPTS` (default: 5) times, and if they fail, it will mark deploy as failed and reject it.
> 
> These also act as smoke tests and saved us numerous times from bad deploys. But they also ensure that containers are really ready to accept traffic. I can probably devote some time to try to implement it in CapRover, but maybe you can give me some hints as how this could be implemented?

i'm also experiencing this issue. There's no zero downtime deployments.

**Comment by activenode** [comment]
> TLDR: CapRover does support zero downtime (mostly). What makes you think that CapRover doesn't have zero downtime?
> 
> More details: If you're app doesn't use a volume, CapRover uses `start-first` strategy when updating the containers. This means that the newer version of your container will be up and running before the old one is killed. This should give you next-to-zero downtime. There is still a slight chance that your container wakes up but internally it takes some time before it can accept requests. This is out of CapRover's hands.
> 
> This strategy is intentionally not applied to apps with volumes attached to them. This is because if multiple instances of the same service are trying to access the same file, this will result in data corruption and failures. For apps with volumes (persistent data), CapRover user `stop-first` strategy. This means that the old container is stopped first, before the new container starts up. This results in a few seconds of downtime.

I think this is a decision you should leave up to the developers as they know their apps best.

See, reason is: With this reasoning you could even argue the same thing without having a volume e.g. If I am using a static build that is using a database elsewhere. You cannot protect devs by assumptions really.

Like e.g. if my volume is used for cdn reasons and my CDN file names are controlled by a database, there is no reason not to let multiple containers access that volume at the same time.

So maybe go rather like having a checkmark:

"You have a volume attached, do you wanna have zerodowntime deployment with the following implications? ..."

**Comment by githubsaturn** [comment]
@activenode  - you are certainly correct and the decision is definitely up to the users. These are simply the default behavior. You can override this (and many other defaults) by using [service override](https://caprover.com/docs/service-update-override.html).

In this case, simply add the following snippet to your service override:
```yaml
UpdateConfig:
  Order: start-first
```

**Comment by activenode** [comment]
Wonderful, thanks @githubsaturn

**Comment by gschurck** [comment]
By using docker HEALTHCHECK and delay option it should be possible to force a zero downtime deployment even for a stateless app with only one replica right ?

**Comment by activenode** [comment]
> By using docker HEALTHCHECK and delay option it should be possible to force a zero downtime deployment even for a stateless app with only one replica right ?

I remember having tried that and I remember it worked, yup.

**Comment by slava-vishnyakov** [comment]
In case someone Googles this for zero-downtime deploys and tries to find a solution that checks _only that container has started_, but then "stops" the check, add this to your Dockerfile

> Oh and I forgot to mention one more benefit that Dokku's zero-downtime deploys have over `HEALTHCHECKS` in CapRover. If you deploy and your HEALTHCHECK for some reason fails, you see that you deployed successful, but you actually see the old version (so, in a sense, your deploy failed) without any notification that your deploy failed.
> 
> I'm not sure how to solve this in CapRover (with the Swarm and all), so I'm just thinking out loud.

Well, a year later, this is a way to do the "deploy" check, not "liveness":

```Dockerfile
HEALTHCHECK --interval=1s --timeout=1s --start-period=1s --retries=30 \
  CMD [ "/bin/sh", "-c", "if [ -f /tmp/started ]; then exit 0; else curl -f http://localhost/ | grep 'ALIVE' && touch /tmp/started || exit 1; fi" ]
```

This tries to connect to `http://localhost/` and if it sees a certain string (`ALIVE` in the example), it stops the checks. It will try to repeat the check `30` times with an interval of `1` second. If successful, it will create `/tmp/started` file and won't hammer `http://localhost/` anymore

**Comment by avarayr** [comment]
@slava-vishnyakov Thank you for that, that is perfect! 
I've modified the script for my use-case where `curl` is not available (Alpine linux)

```Dockerfile
HEALTHCHECK --interval=1s --timeout=1s --start-period=1s --retries=30 \
  CMD [ "/bin/sh", "-c", "if [ -f /tmp/started ]; then exit 0; else wget -qO- `hostname -i` | grep -q '__next' && touch /tmp/started || exit 1; fi" ]
```
- Uses wget which is available on most distros
- Resolves the healthcheck ping domain by `hostname -i` instead of `localhost`, because localhost is not available inside caprover apps.
- Checks for `__next` in the health-check endpoint response (my usecase is a next.js app), will need to modify for other apps

**Comment by Boscop** [comment]
@githubsaturn 
> This strategy is intentionally not applied to apps with volumes attached to them. This is because if multiple instances of the same service are trying to access the same file, this will result in data corruption and failures. 

What about apps with volumes mounted as read-only, will CapRover use a `start-first` strategy and thus support zero-downtime deployments for these?

**Comment by githubsaturn** [comment]
Good question! CapRover volumes (the ones you define in the UI) aren't read only. But if you manually set up a volume using service override, you can definitely set it up as a read only volume. And in that case, in the same service override, you can define start-first as your update strategy. 

It's really as simple as having this in your service override:
```yaml
UpdateConfig:
  Parallelism: 2
  Delay: 1000000000
  FailureAction: pause
  Monitor: 15000000000
  MaxFailureRatio: 0.15
  Order: start-first
```

**Comment by LaundroMat** [comment]
I have a similar problem.

My Dockerfile has an ENTRYPOINT that takes about a minute to run (database migrations, etc). 

When I deploy to caprover, the app is unreachable during that period.

I've tried the following, but none have helped:

* Increase the number of instances;
* Add HEALTHCHECK to the end of the Dockerfile
```
HEALTHCHECK --interval=2s --timeout=5m --start-period=3s CMD wget `hostname -i`:8000 || exit 1
ENTRYPOINT [ "bash", "-c", "./entrypoint.sh" ]
EXPOSE 8000
```
* Override the service config to 
```
UpdateConfig:
  Order: start-first
```

Between the moment of the succesful build and the actual app server going live, there is about a minute in which the site is unreachable (502).

**Comment by githubsaturn** [comment]
You don't need to increase number of instance. Just HEALTHCHECK should be enough.

I tried this and it works fine. Here is an example repository in case you are interested: 
https://github.com/githubsaturn/zero-downtime

This repository is just a simple nginx image, but it has a 30sec countdown on entrypoint. 

You can try deploying it on your instance and you'll see that it just works. 
* Create a new app (**no persistent data**)
* Go to deployment tab and enter these values, then SAVE & UPDATE
* Finally click on FORCE BUILD.
* You will see that after the build is finished, the old place holder is being displayed for 30sec WITHOUT any 502 errors. But after 30sec, the app switches to the new image.

I've deployed on demo instance so you can see it for yourself. (password: captain42)
https://captain.server.demo.caprover.com

App name: `test-healthcheck`

**Comment by githubsaturn** [comment]
A new article is now added to the documentation to help better understand this issue:
https://caprover.com/docs/zero-downtime.html

**Comment by LaundroMat** [comment]
It works, but I have another problem now: entrypoint.sh is being executed in an infinite loop.
When I remove the healthcheck, this doesn't happen, so it's not a problem with the code in the entrypoint.sh file.
The final lines of my Dockerfile read:

```

RUN ["chmod", "+x", "./entrypoint.sh"]

HEALTHCHECK --interval=3s --timeout=1s --retries=15 \
  CMD curl -f http://0.0.0.0:8000 || exit 1

ENTRYPOINT [ "bash", "-c", "./entrypoint.sh" ]
# ENTRYPOINT ["/code/entrypoint.sh"]

EXPOSE 8000

```

I have to test for 0.0.0.0:8000 because that's how gunicorn serves my Django app.
